### PR TITLE
[JENKINS-45842] - Provide minimal Java level in Jenkins API

### PR DESCRIPTION
**Pull request by oleg-nenashev**
See [JENKINS-45842](https://issues.jenkins-ci.org/browse/JENKINS-45842). The idea is to make it possible to determine the required Java versions without hardcoding them everywhere. See [JENKINS-43500](https://issues.jenkins-ci.org/browse/JENKINS-43500) for issue examples. Downstream PR: https://github.com/jenkinsci/ssh-slaves-plugin/pull/66

I also added an option to override the requirement via a system property

I will create a downstream PR in the SSH plugin soon.

### Proposed changelog entries

* Entry 1: Issue, Human-readable Text
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @daniel-beck

### CODE REVIEW

File: core/src/filter/resources/jenkins/model/jenkins-version.properties
Comment (recena) [inline core/src/filter/resources/jenkins/model/jenkins-version.properties]: @oleg-nenashev What do you think if we put this value in the `MANIFEST`?

Comment (oleg-nenashev) [inline core/src/filter/resources/jenkins/model/jenkins-version.properties]: We could do it. I just used the existing infrastructure

Comment (recena) [inline core/src/filter/resources/jenkins/model/jenkins-version.properties]: @oleg-nenashev If I was my choice I would remove `jenkins-version.properties`. IMHO the right place is the standard file: `MANIFEST`.

I'm not sure if `jenkins-version.properties` was created for including other properties, at least, the name does not tell that.

Comment (oleg-nenashev) [inline core/src/filter/resources/jenkins/model/jenkins-version.properties]: The file is a generic property file, so it should not really matter how many properties are there. I would also rework the logic, but I do not think it's something to be done in this PR. WDYT?

Comment (recena) [inline core/src/filter/resources/jenkins/model/jenkins-version.properties]: @oleg-nenashev it is property file, but its name shows / tells a purpose. If you find a `mysql.properties` you expect to see _things_ related with MySQL, for example, a configuration, etc...

My suggestion would be to add a new entry in the `MANIFEST` and read from there the value.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: I don't understand what's going on here. Is all of this to support a custom (higher) minimum release is repackaged Jenkins distributions?

**Comment by daniel-beck** [comment]
The issue description reads as if plugins could contribute to the minimum level somehow, possibly increasing it if compiled against newer Java. That's not in scope for this change though, is it?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: > I don't understand what's going on here. Is all of this to support a custom (higher) minimum release is repackaged Jenkins distributions?

AFAIK there is no custom distributions requiring it, but plugin maintainers may bump Java level beyond the core. We had it for Yet Another Docker plugin and several else. If somebody uses Java 9 plugins in the future, they can mark the requirement as a system property

**Comment by oleg-nenashev** [comment]
> The issue description reads as if plugins could contribute to the minimum level somehow, possibly increasing it if compiled against newer Java. That's not in scope for this change though, is it?

@daniel-beck I had no plan to do that, but I could add it. E.g. by offering a new method in `Plugin` or a new extension point.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: > they can mark the requirement as a system property

There must be a better approach than that for this, ideally something integrating with plugin metadata, e.g. defined properties in the POM, or manifest entries, so we can publish PRE requirements via update center -- I think there was a change to maven-hpi-plugin proposed a while back.

"Install (or update) a plugin and see Jenkins not starting up again" seems like not a great experience.

Comment (alexanderrtaylor) [inline core/src/main/java/jenkins/model/Jenkins.java]: @oleg-nenashev this does not stop Jenkins from starting right? This just prints a log which says your Java version is outdated correct?

Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: @alexanderrtaylor Jenkins won't start up due to https://github.com/jenkinsci/extras-executable-war/pull/8/ . This change just adds new API which can be used to determine the current requirement. For plugins the behavior is TBD, but it has to be done on the update center side. Once the plugin with custom Java req is installed, you cannot do much excepting failing horribly on startup

**Comment by oleg-nenashev** [comment]
@reviewbybees ping.

**Comment by daniel-beck** [comment]
@oleg-nenashev Didn't we determine this is an unworkable solution once we take update center metadata into account? What exactly are you asking for here?

**Comment by oleg-nenashev** [comment]
@daniel-beck 

> Didn't we determine this is an unworkable solution once we take update center metadata into account? 

IIRC in IRC we agreed that we can ship it in the current state and then extend the implementation to take plugins into account (and keep the same API).

Plugin support was not in the scope of the JIRA ticket. I commented there to make it explicit.

**Comment by daniel-beck** [comment]
@oleg-nenashev We agreed that ssh-slaves can be shipped without the core API addition.

> Plugin support

So what's all of this even for, then? Who's supposed to set the system property, and why? In fact, you wrote in a comment earlier:

> AFAIK there is no custom distributions requiring it, but plugin maintainers may bump Java level beyond the core. We had it for Yet Another Docker plugin and several else. If somebody uses Java 9 plugins in the future, they can mark the requirement as a system property

If this is not for use by plugins (which may result in Jenkins failing to launch!), what is it for?

**Comment by daniel-beck** [comment]
I discussed this directly with @oleg-nenashev and he pointed out that admins may wish to set the system property as they know they might be running plugins incompatible with lower Java releases (e.g. 9 on an otherwise 8 core), and then any plugins querying the required Java release should return that version. So from that perspective, system property etc. make sense.

**Review by alexanderrtaylor — APPROVED** [review]
🐝 It seems like this could potentially cause confusion when a user is not properly reading plugin changes but the code looks good and I certainly see a use case.

**Comment by oleg-nenashev** [comment]
@alexanderrtaylor I am not sure I understand your comment. This code is not impact the plugin changes UX, nothing really becomes worse or better there.

**Comment by alexanderrtaylor** [comment]
@oleg-nenashev Sorry I was just trying to imagine a user experience using Java 7 where they go to update a plugin which now requires Java 8. Then when the plugin is updated properly and they restart Jenkins does not come up until either they disable the plugin or install Java 8.

That would be the experience right? I just worry about people not reading changelogs when a plugin now requires newer Java versions.

**Comment by daniel-beck** [comment]
@alexanderrtaylor This addresses a different issue. I had the same misconception about this change originally. This change does the following:

* Add an API that lets plugins query the version of Java required by Jenkins. This API could `return 8` but instead uses a system property so that…
* admins can declare a higher required Java dependency because they know that their instance Jenkins does not work with the core minimum.

This will let e.g. `ssh-slaves` select a different JRE to run the agent with.

A proper declarative way (ideally integrated with `PluginManager`, backend-update-center2, and possibly maven-hpi-plugin) for plugins to specify what JRE they minimally require is the next step and not implemented here. Plugins are not expected to override this system property to require a newer JRE.

**Comment by alexanderrtaylor** [comment]
@oleg-nenashev @daniel-beck Ahh OK then I misunderstood what this API was doing. Thanks for the clarification and I take my comments back :)

**Comment by oleg-nenashev** [comment]
@reviewbybees done 
It needs more reviews though.

**Review by ghost — APPROVED** [review]
:bee:

File: core/src/filter/resources/jenkins/model/jenkins-version.properties
Comment (jtnord) [inline core/src/filter/resources/jenkins/model/jenkins-version.properties]: properties should not start with "`java.`" as that namespace should be considered reserved for properties coming from java itself (to avoid any future blow up when Java introduces a new property)

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: > This change just adds new API which can be used to determine the current requirement

But if that is not met how can you get it (the class format will not be compatable)?   if you can get it - you don't care as the requirement must have been met...

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: And if a plugin is installed that uses Java9?  if you expect this to then return Java9 then what if that plugin only executes code on the master and never on an agent?  Why would that then force callers (the downstream ssh PR) to use a higher agent version? unnesceserily.  At this point I am not sure what this is trying to acheive - and no property is needed as this can be found out at runtime as you have all the information you need in the class file.

File: core/src/filter/resources/jenkins/model/jenkins-version.properties
Comment (oleg-nenashev) [reply core/src/filter/resources/jenkins/model/jenkins-version.properties]: Would "jenkins.java.min.level" work for you?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/model/Jenkins.java]: > And if a plugin is installed that uses Java9? if you expect this to then return Java9 then what if that plugin only executes code on the master and never on an agent

We consider Jenkins to be a shared context platform. If the class is not annotated as `@OnMaster`, it may be actually sent to the agent by another class. It's hard to control that. As a happy Remoting maintainer, I would be against supporting different Java versions on master and agents.

We could add that (just via 2 properties), but IMHO it needs a real use-case. And, even there is such case, the support can be added later in a compatible way.

**Comment by oleg-nenashev** [comment]
merge conflict.
@jtnord i am waiting for your response

File: core/src/filter/resources/jenkins/model/jenkins-version.properties
Comment (jtnord) [reply core/src/filter/resources/jenkins/model/jenkins-version.properties]: it would work but "min and level" are not really hierachical - so `jenkins.java.minLevel` or `jenkins.java.minVersion` would be preferable in my view.

**Comment by jtnord** [comment]
@daniel-beck wrote
> This will let e.g. ssh-slaves select a different JRE to run the agent with.

But then We don;t want to use that in SSH agent - as @oleg-nenashev wrote..

> We consider Jenkins to be a shared context platform. ... As a happy Remoting maintainer, I would be against supporting different Java versions on master and agents.

and as a user using ssh agents you would want them to use the same java version that runs on the master and not the minimum version that Jenkins (and plugins?) supports.

So I am kinda confised as I wrote above

> This change just adds new API which can be used to determine the current requirement
But if that is not met how can you get it (the class format will not be compatable)? if you can get it - you don't care as the requirement must have been met...

**Comment by oleg-nenashev** [comment]
Resolved the merge conflict. Will consider finalizing it during the Java 10+ hackathon if time allows

**Comment by oleg-nenashev** [comment]
@jenkinsci/java11-support is it something interesting in the scope of the Java 11 work?

**Comment by oleg-nenashev** [comment]
@jenkinsci/java11-support ping about opinions

**Comment by batmat** [comment]
Not sure I get the scope/goal for this TBH.

**Comment by daniel-beck** [comment]
I don't understand how https://github.com/jenkinsci/ssh-slaves-plugin/pull/66 is downstream but merged and released long ago. There does not appear to be a real downstream plugin change that would demonstrate how this would actually be used in plugins.

**Comment by rsandell** [comment]
LGTM but might be outdated and will need a proper changelog entry

**Comment by alecharp** [comment]
We might also need to trigger a build on this PR and see if it needs to get `master` latest changesets.

**Comment by MRamonLeon** [comment]
@oleg-nenashev could you please rerun the CI?

**Comment by varyvol** [comment]
@oleg-nenashev do you think this still would add some value? We've gone through JDK11 changes without this.

**Comment by MRamonLeon** [comment]
It seems there is no consensus on the approach followed here. Maybe we can discuss how to solve the problem in the Jenkins dev-list before filing the PR to fix it. Do you agree on that @oleg-nenashev? Thanks

**Comment by rsandell** [comment]
Closing because no activity. You are welcome to reopen when work is resumed.

**Pull request closed without merging**
